includes/class-wc-taxjar-api-calculation.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+class WC_Taxjar_API_Calculation {
+
+	/**
+	 * @var WC_Taxjar_Integration
+	 */
+	public $taxjar_integration;
+
+	/**
+	 * WC_Taxjar_API_Calculation constructor.
+	 */
+	public function __construct( $integration ) {
+		$this->taxjar_integration = $integration;
+
+		if ( $this->is_api_calculation_enabled() ) {
+			// Calculate tax during creation and update of order through WooCommerce REST API
+			add_filter( 'woocommerce_rest_pre_insert_shop_order_object', array( $this, 'calculate_api_order_tax'), 20, 3);
+		}
+	}
+
+	/**
+	 * Determine is tax calculation on API orders is enabled
+	 *
+	 * @return bool
+	 */
+	public function is_api_calculation_enabled() {
+		return isset( $this->taxjar_integration->settings['api_calcs_enabled'] ) && 'yes' == $this->taxjar_integration->settings['api_calcs_enabled'];
+	}
+
+	/**
+	 * Calculates tax on order created through the API
+	 *
+	 * @param WC_Order $order Object object.
+	 * @param WP_REST_Request $request Request object.
+	 * @param bool $creating If is creating a new object.
+	 *
+	 * @return WC_Order
+	 */
+	public function calculate_api_order_tax( $order, $request, $creating ) {
+
+		if ( ! $this->api_order_needs_tax_calculated( $order, $request, $creating ) ) {
+			return $order;
+		}
+
+		$this->taxjar_integration->calculate_order_tax( $order );
+
+		return $order;
+	}
+
+	/**
+	 * Determines whether or not to calculate tax on and API order
+	 *
+	 * @param WC_Order $order Object object.
+	 * @param WP_REST_Request $request Request object.
+	 * @param bool $creating If is creating a new object.
+	 *
+	 * @return bool
+	 */
+	public function api_order_needs_tax_calculated( $order, $request, $creating ) {
+		$needs_tax_calculated = true;
+
+		if ( ! $creating ) {
+			if ( ! isset( $request['billing'] ) && ! isset( $request['shipping'] ) && ! isset( $request['line_items'] ) && ! isset( $request['shipping_lines'] ) && ! isset( $request['fee_lines'] ) && ! isset( $request['coupon_lines'] ) ) {
```

Comment:
@dallendalton Can you briefly explain the logic here and why it's done this way in WooCommerce core?

Reply:
Basically for any order update request sent to the Woo REST API, if the request contains data that would affect the totals calculation, WooCommerce will already recalculate tax for the order (as well as all the other totals). So this includes any change to billing, shipping, or any of the various types of line items (fees, products, shipping and coupons). By mimicking this logic we are basically saying that when woo has decided to recalculate tax it will be done through TaxJar.